File: python/src/birs/birs.py
from functools import reduce
from src.birs.libs import basic_influence_role, sort_data, roles_frequencies


class BIRs:
    """Basic Influence Roles"""

    @staticmethod
    def rank(data):
        """Add ranking score"""
        data.sort(key=sort_data, reverse=True)
        for i, n in enumerate(data): n['rank'] = i + 1
        return data

    @staticmethod
    def detect(indegree, outdegree, node_count, data=False):
        """Detect Basic Influence Role """
        return basic_influence_role(indegree, outdegree, node_count, data)

    @staticmethod
    def distribution(data):
        """Calculate distribution of BIRs"""
        node_count = len(data)
        agg_roles = reduce(roles_frequencies, data, {})
        for _, vr in agg_roles.items():
            vr.update(dict(
                frequency=vr['count'] / node_count,
                levels={
                k: dict(count=v, frequency=v / node_count)
                for k, v in vr['levels'].items()
            }))
        return agg_roles

    @staticmethod
    def detect_from_nodes(nodes):
        """Given a list of nodes and their indegree and outdegree, return their BIRs"""
        node_count = len(nodes)
        data = [
            dict(id=n['id'], **BIRs.detect(n['indegree'], n['outdegree'], node_count, data=True))
            for n in nodes
        ]
        return BIRs.rank(data)

    @staticmethod
    def detect_nx(G):
        """Given a NetworkX DiGraph, return BIRs of nodes"""
        if not G.is_directed():
            raise Exception('The graph must be directed: undirected networks are not supported.')
        node_count = G.number_of_nodes()
        data = [
            dict(id=n, **BIRs.detect(G.in_degree(n), G.out_degree(n), node_count, data=True))
            for n in G.nodes()
        ]
        return BIRs.rank(data)

File: javascript/libs/birs.js
'use strict';

const {basicInfluenceRole, sortData, rolesFrequencies } = require('./libs');

// Add ranking score
const rank = data => data.sort(sortData).reverse().map((d, i) => { d.rank = i + 1; return d; });

// Detect Basic Influence Role
const detect = (indegree, outdegree, nodeCount, data) => {
    if (indegree >= nodeCount || outdegree >= nodeCount) throw Error('Node count must be greater than indegree or outdegree.');
    return basicInfluenceRole(indegree, outdegree, nodeCount, data);
}

const distribution = data => {
    // Calculate distribution of BIRs
    const nodeCount = data.length;
    const aggRoles = data.reduce(rolesFrequencies, {});
    for (const role in aggRoles) {
        const vr = aggRoles[role];
        vr.frequency = vr.count / nodeCount;
        vr.levels = Object.entries(vr.levels).reduce((acc, [level, count]) => {
            acc[level] = { count, frequency: count / nodeCount };
            return acc;
        }, {});
    }
    return aggRoles;
}

const detectFromNodes = nodes => {
    // Given a list of nodes and their indegree and outdegree, return their BIRs
    const nodeCount = nodes.length;
    const data = nodes.map(n => ({
        id: n.id,
        ...detect(n.indegree, n.outdegree, nodeCount, true),
    }));
    return rank(data);
}

const detectGraphology = G => {
    // Given a Graphology DiGraph, return BIRs of nodes
    if (G.type != 'directed') throw new Error('The graph must be directed: undirected networks are not supported.');
    const nodeCount = G.order;
    const data = G.mapNodes(n => ({
        id: n,
        ...detect(G.inDegree(n), G.outDegree(n), nodeCount, true),
    }));
    return rank(data);
}

module.exports = {
    detect,
    detectFromNodes,
    detectGraphology,
    distribution
}
  